### DISCUSSION TITLE
Recursive CTE has wrong number of columns when combined with joined table inheritance

### DISCUSSION THREAD

**Discussion by lamflokas** [sqlalchemy/sqlalchemy#10123] [Usage Questions] [answered]
Hello everyone! I am having some problems with running a query involving a recursive CTE that also involve joined table inheritance. My application use case is much more complicated, but I boiled it down to the fact that the WITH RECURSIVE statement generated is incorrect. Here is a minimal example that reproduces the issue in SQLAlchemy 2.0.17:

```python
from sqlalchemy import ForeignKeyConstraint
from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session

engine = create_engine('sqlite:///:memory:')

class Base(DeclarativeBase):
    pass

class Employee(Base):
    __tablename__ = "Employee"

    id: Mapped[str] = mapped_column(primary_key=True, index=True)
    discriminator: Mapped[str] = mapped_column()
    general : Mapped[str] = mapped_column()

    __mapper_args__ = {"polymorphic_on": discriminator, "polymorphic_abstract": True}

class Manager(Employee):
    __tablename__ = "Manager"
    
    id: Mapped[str] = mapped_column(primary_key=True, index=True)
    special : Mapped[str] = mapped_column()

    __mapper_args__ = {
        "polymorphic_identity": "Manager",
    }

    __table_args__ = (
        ForeignKeyConstraint(
            ["id"],
            [
                "Employee.id"
            ],
        ),
    )    

Base.metadata.create_all(engine)

my_cte = (select(Manager)
          .filter(Manager.special == 'my special')
          .filter(Manager.general == 'my general')
          .cte(recursive=True))

session = Session(engine)
session.scalars(select(my_cte)).all()
```

The important part of the error is the following
```
OperationalError: (sqlite3.OperationalError) table anon_1 has 5 values for 4 columns
[SQL: WITH RECURSIVE anon_1(id, discriminator, general, special) AS 
(SELECT "Manager".id AS id, "Employee".id AS id_1, "Employee".discriminator AS discriminator, "Employee".general AS general, "Manager".special AS special 
FROM "Employee" JOIN "Manager" ON "Employee".id = "Manager".id 
WHERE "Manager".special = ? AND "Employee".general = ?)
 SELECT anon_1.id, anon_1.discriminator, anon_1.general, anon_1.special 
FROM anon_1]
[parameters: ('my special', 'my general')]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```

You can see clearly that the WITH RECURSIVE statement is incorrect. The schema declaration part correctly specifies that the result has 4 columns which is what I expected. But the body of the WITH RECURSIVE statement is duplicating the id column (one from Manager and one from Employee). The main SELECT that follows looks good to me.

It seems to me that this is a SQLAlchemy bug because with a little help SQLAlchemy can run the query. Here is the workaround that I have identified for this example:

```python
from sqlalchemy import ForeignKeyConstraint
from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session
from sqlalchemy.orm import aliased

engine = create_engine('sqlite:///:memory:')

class Base(DeclarativeBase):
    pass

class Employee(Base):
    __tablename__ = "Employee"

    id: Mapped[str] = mapped_column(primary_key=True, index=True)
    discriminator: Mapped[str] = mapped_column()
    general : Mapped[str] = mapped_column()

    __mapper_args__ = {"polymorphic_on": discriminator, "polymorphic_abstract": True}

class Manager(Employee):
    __tablename__ = "Manager"
    
    id: Mapped[str] = mapped_column(primary_key=True, index=True)
    special : Mapped[str] = mapped_column()

    __mapper_args__ = {
        "polymorphic_identity": "Manager",
    }

    __table_args__ = (
        ForeignKeyConstraint(
            ["id"],
            [
                "Employee.id"
            ],
        ),
    )    

Base.metadata.create_all(engine)

my_cte = (select(Manager.id, Manager.special, Manager.general, Manager.discriminator).
          filter(Manager.special == 'my special').
          filter(Manager.general == 'my general').
          cte(recursive=True))
my_alias = aliased(Manager, my_cte)
session = Session(engine)
session.scalars(select(my_alias)).all()
```

This workaround gives an empty list as expected.

I wanted to ask the community if this is indeed a bug or if I am misusing joined table inheritance, recursive CTEs or both. Do you see any issues with my proposed workaround if applied in more complicated settings?  

EDIT 1: Above I am claiming that the WITH RECURSIVE schema is the correct one and that its body is the incorrect one. But I then discovered that the select on Manager always gives the duplicated primary key:
```python
print(select(Manager))
```
Prints as follows:
```
SELECT "Manager".id, "Employee".id AS id_1, "Employee".discriminator, "Employee".general, "Manager".special 
FROM "Employee" JOIN "Manager" ON "Employee".id = "Manager".id
```
But I guess the second id column does not generally seem to be a problem in simpler queries. I find this rather surprising. I would expect the query to have only one copy of the primary key columns. Is this the intended behavior?

EDIT 2: In my effort to generalize my previous workaround I made some very surprising observations. 
```python
cols = Manager.__mapper__.attrs
print(select(*cols).filter(Manager.special == "my special").filter(Manager.general == "my general"))
```
results in a query without a join between Manager and Employee
```
SELECT "Manager".id, "Employee".discriminator, "Employee".general, "Manager".special 
FROM "Manager", "Employee" 
WHERE "Manager".special = :special_1 AND "Employee".general = :general_1
```
But going through instrumented attributes generates the correct join clause
```python
cols = [x for x in Manager.__dict__.values() if isinstance(x, sqlalchemy.orm.InstrumentedAttribute)]
print(select(*cols).filter(Manager.special == "my special").filter(Manager.general == "my general"))
```
printing
```
SELECT "Manager".id, "Manager".special, "Employee".discriminator, "Employee".general 
FROM "Employee" JOIN "Manager" ON "Employee".id = "Manager".id 
WHERE "Manager".special = :special_1 AND "Employee".general = :general_1
```
Of course using the columns of the underlying tables directly also does not render the join condition. 
I was wondering if there are some best practices on how to avoid these subtle differences when using joined table inheritance.

Thank you in advance!

**Comment by zzzeek** [comment]
hi -

yes these are likely all bugs.  from a brief scan it's not clear how many individual bugs you've located or if you have workarounds for now, however, if you can isolate each individual "surprising" behavior into distinct issues, we can try to look at them.    there are multiple, unfortunate circumstances however:

1. the CTE logic is **extremely** complicated and was largely contributed by other users.   it's not clear how difficult a bug here involving joined inheritance would be to fix
2. I was just assigned two other very deep, intricate bugs by another also regarding joined inheritance.  each of these bugs looks like a full day of work
3. I am leaving for vacation tomorrow for a week and have only a few hours to look at SQLAlchemy issues today.

I've only skimmed what you are seeing and it is not clear to me that you have achieved workarounds or not for your immediate issue.     

If I can look quickly, I would say the `select(*Manager.__mapper__.attrs)` thing is not surprising, I would not expect that to work; attrs gives you `MapperProperty` objects and these are not appropriate right now as elements to be passed to `select()`.    To the degree there's a "bug" here SQLAlchemy would be better off raising for this use, but it's not clear how hard it would be to detect that condition.  Also, as this falls under the category of "things we never thought of, but other people did and it apparently works to some extent, but not completely, so we have to tell them this, but production code is possibly relying on this", we cant add a raise for that, it would be a warning.     So bug number one (if you could please open separately), "Warn when MapperProperty objects are passed to select()".

From there, the CTE thing is probably a bug as well, but even looking at these issues takes me a few hours which I unfortunately do not have right now, but please make a succinct issue for that one as well, for the single exact thing that's not working.  thanks

**Comment by zzzeek** [comment]
the initial issue has a fix in progress at #10169

**Reply by lamflokas** [reply, marked as answer]
The initial issue is indeed fixed in 2.20. Thank you @zzzeek !
